jena-arq/src/main/java/org/apache/jena/sparql/engine/iterator/QueryIterFailed.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.jena.sparql.engine.iterator;
+
+import org.apache.jena.query.QueryCancelledException;
+import org.apache.jena.query.QueryExecException;
+import org.apache.jena.sparql.engine.ExecutionContext;
+import org.apache.jena.sparql.engine.QueryIterator;
+import org.apache.jena.sparql.engine.binding.Binding;
+
+/**
+ * A helper iterator for when an intermediate step in the construction of a compound iterator fails.
+ */
+public class QueryIterFailed
+    extends QueryIter1
+{
+    private Throwable cause;
+
+    public QueryIterFailed(QueryIterator input, ExecutionContext execCxt, Throwable cause) {
+        super(input, execCxt);
+        this.cause = cause;
+    }
+
+    @Override
+    protected void requestSubCancel() {
+    }
+
+    @Override
+    protected void closeSubIterator() {
+    }
+
+    @Override
+    protected boolean hasNextBinding() {
+        throw buildException();
+    }
+
+    @Override
+    protected Binding moveToNextBinding() {
+        throw buildException();
+    }
+
+    protected RuntimeException buildException() {
```

Comment:
Any reason not to make this a `QueryExecException`?

Reply:
Changed to QueryExecException.